---
layout: page
title: Getting Started
summary: "Guides users on the initial steps, including system requirements, downloading Musta, 
installation instructions, and verifying the installation."
---

---
## System Requirements
To install and run **Musta**, it is essential to have Docker installed on your computer. 
Docker is a containerization platform that ensures consistency and compatibility in running **Musta** 
across different computing environments.

If you do not have Docker installed, 
you can download and install it by following the instructions provided in the official Docker documentation:
[Docker Installation Guide](https://docs.docker.com/engine/install).

Please ensure that Docker is properly configured and running on your system before proceeding with the installation of **Musta**.

---
## Downloading Musta
To obtain **Musta**, you can clone the official **Musta** repository using Git. 
Open a terminal or command prompt and use the following Git command:

```git
git clone https://github.com/next-crs4/musta.git
```
This command will download the latest version of Musta to your local system. 
Once the repository is cloned, 
you can proceed with the installation and configuration steps to set up **Musta** for your somatic mutation analysis tasks.

---
## Installation Instructions

The installation could require several minutes

1. **Change Directory to Musta**: After cloning the **Musta** repository from GitHub, navigate to the Musta directory using the cd command:
```shell
    cd musta
```
2. **Bootstrap the Musta Framework:** To set up the **Musta** framework, use the following command:
```shell
  make bootstrap
```
    This command will initiate the installation and configuration process, ensuring that Musta is ready for use.

<pre style="font-size:12px">[+] Building 384.6s (20/20) FINISHED                                                                                                                                                                                                
<font color="#268BD2"> =&gt; [internal] load build definition from Dockerfile                                                           0.1s</font>
<font color="#268BD2"> =&gt; =&gt; transferring dockerfile: 1.37kB                                                                         0.0s</font>
<font color="#268BD2"> =&gt; [internal] load .dockerignore                                                                              0.1s</font>
<font color="#268BD2"> =&gt; =&gt; transferring context: 2B                                                                                0.0s</font>
<font color="#268BD2"> =&gt; [internal] load metadata for docker.io/library/python:3.8                                                  0.0s</font>
<font color="#268BD2"> =&gt; [internal] load build context                                                                              0.2s</font>
<font color="#268BD2"> =&gt; =&gt; transferring context: 1.18MB                                                                            0.0s</font>
<font color="#268BD2"> =&gt; [ 1/15] FROM docker.io/library/python:3.8                                                                  0.3s</font>
<font color="#268BD2"> =&gt; [ 2/15] RUN mkdir /code                                                                                    0.2s</font>
<font color="#268BD2"> =&gt; [ 3/15] WORKDIR /code                                                                                      0.1s</font>
<font color="#268BD2"> =&gt; [ 4/15] COPY . /code                                                                                       0.1s</font>
<font color="#268BD2"> =&gt; [ 5/15] RUN mkdir /config                                                                                  0.4s</font>
<font color="#268BD2"> =&gt; [ 6/15] ADD /config/* /config/                                                                             0.1s</font>
<font color="#268BD2"> =&gt; [ 7/15] RUN apt-get -qq update &amp;&amp; 
apt-get install --no-install-recommends -y dialog apt-utils software-properties-common git wget curl bzip2 &amp;&amp;
apt-get autoremove -y &amp;&amp; apt-get clean &amp;&amp; rm -rf /var/lib                                                     18.4s</font>
<font color="#268BD2"> =&gt; [ 8/15] RUN groupadd -g 1000 appuser &amp;&amp;     useradd -m -u 1000 -g appuser appuser                          0.5s</font>
<font color="#268BD2"> =&gt; [ 9/15] RUN curl -L https://repo.anaconda.com/miniconda/Miniconda3-py38_4.12.0-Linux-x86_64.sh &gt; 
miniconda.sh &amp;&amp; sh miniconda.sh -b -p /opt/conda &amp;&amp; rm miniconda.sh                                           16.2s</font> 
<font color="#268BD2"> =&gt; [10/15] RUN conda update -n base -c defaults conda                                                        54.1s</font> 
<font color="#268BD2"> =&gt; [11/15] RUN conda config --set channel_priority strict                                                     0.6s</font> 
<font color="#268BD2"> =&gt; [12/15] RUN  conda create -q -y -c conda-forge -c bioconda -n musta  python=3.8 snakemake=7.15 &amp;&amp;
conda clean --all -y                                                                                         282.0s</font> 
<font color="#268BD2"> =&gt; [13/15] RUN sh /config/create_paths.sh                                                                     0.4s</font> 
<font color="#268BD2"> =&gt; [14/15] RUN cd /code/src &amp;&amp; make install &amp;&amp; cd /code                                                       4.7s</font> 
<font color="#268BD2"> =&gt; [15/15] RUN echo &quot;source activate musta&quot; &gt; ~/.bashrc                                                       0.4s</font> 
<font color="#268BD2"> =&gt; exporting to image                                                                                         5.7s</font> 
<font color="#268BD2"> =&gt; =&gt; exporting layers                                                                                        5.7s</font> 
<font color="#268BD2"> =&gt; =&gt; writing image sha256:bb170cfc65469c4a3445a04c3b21c843c5151e84ac8276107323325da973c901                   0.0s</font> 
<font color="#268BD2"> =&gt; =&gt; naming to docker.io/library/musta:Dockerfile                                                            0.0s</font> 
                                                                                                                                                                                                                                    
Ready to start. Try:
	musta --help
</pre>

Following these steps, you will have successfully installed **Musta** on your system, 
and it will be ready for somatic mutation analysis. You can now proceed to configure and use **Musta** for your specific research needs.

---

## Verifying the Installation
To confirm that the Musta Docker image has been successfully built, you can use the `docker images` command. 
Open a terminal or command prompt and enter the following command:
```shell
docker images
```
After running this command, you should see an output that resembles the following:
```text
REPOSITORY   TAG          IMAGE ID         CREATED       SIZE
musta        Dockerfile   bb170cfc6546     2 hours ago   2.48GB
```
The output should display the "musta" repository, "Dockerfile" tag, image ID, creation date, and image size. 
If you see similar information in your output,
 it indicates that the Musta Docker image has been successfully built and is ready for use in your environment.

To confirm that the Musta Command Line Interface (CLI) is correctly installed, 
you can use the `musta --help` command. Open a terminal or command prompt and enter the following command:

```shell
musta --help
```
When you run this command, you should see an output similar to the following:

```text
usage: musta [-h] [--config_file PATH] [--logfile PATH]
             [--loglevel {DEBUG,INFO,WARNING,ERROR,CRITICAL}]
             {detect,classify,interpret,end2end} ...

End-to-end pipeline to detect, classify and interpret mutations in cancer

optional arguments:
  -h, --help            show this help message and exit
  --config_file PATH, -c PATH
                        configuration file
  --logfile PATH        log file (default=stderr)
  --loglevel {DEBUG,INFO,WARNING,ERROR,CRITICAL}
                        logger level.

subcommands:
  valid subcommands

  {detect,classify,interpret,end2end}
                        sub-command description
    detect              Somatic Mutations Detection.
                            1.  Multiple Variant Calling: mutect, lofreq, varscan, 
                                vardict, muse, strelka.
                            2.  Ensemble consensus approach to combine results and 
                                to improve the performance of variant calling
    classify            Variant Annotation
                        Functional annotation of called somatic variants 
                        
    interpret           Somatic Mutations Interpretation:
                            1.  Identification of cancer driver genes 
                            2.  Check for enrichment of known oncogenic pathways.
                            3.  Infer tumor clonality by clustering variant allele frequencies.
                            4.  Deconvolution of Mutational Signatures
                            
    end2end             Run the whole workflow, 
                        from Variant Calling to Deconvolution of Mutational Signatures

```

If you receive this output, 
it indicates that the **Musta CLI** is correctly installed and ready for use. 
You can now proceed with using **Musta** for somatic mutation analysis in cancer research.

---

## Basic Usage

The Musta command structure is as follows:

```shell
musta COMMAND --workdir WORKING-DIR --samples-file SAMPLES-FILE [options]
```

Here's an overview of these components:

* **--workdir WORKING-DIR**: Designates the destination folder for analysis, where the Snakemake pipeline, logs, and analysis outputs are located.  
* **--samples-file SAMPLES-FILE**: Points to a YAML file listing the datasets you wish to analyze.
* **COMMAND**:
  1. `detect`: Somatic Mutations Detection.
     - Multiple Variant Calling: Mutect, Lofreq, Varscan, Vardict, Muse, Strelka.
     - Ensemble consensus approach to combine results and improve the performance of variant calling.
  2. `classify`: Variant Annotation.
     - Functional annotation of called somatic variants.
  3. `interpret`: Somatic Mutations Interpretation.
     - Identification of cancer driver genes. 
     - Checking for the enrichment of known oncogenic pathways. 
     - Inferring tumor clonality by clustering variant allele frequencies. 
     - Deconvolution of Mutational Signatures.
  4. `end2end`: Run the whole workflow, from Variant Calling to Deconvolution of Mutational Signatures.

---

## Removing Musta
To remove **Musta** from your system, you can use the following command:

```shell
make clean
```

This command will clean up and remove any installed **Musta** components, ensuring that **Musta** is no longer present on your system.

<p style="text-align:right"><a href="{{site.url}}{{page.url}}"><strong>Go Up</strong><span class="fa fa-fw fa-arrow-up"></span></a></p>
---