---
layout: page
title: Classification
summary: "Users learn how to annotate somatic mutations and understand the significance of functional annotations. 
It also covers using different variant annotators."
---

--- 
Following the variant calling process, where somatic mutations are identified, it is essential \
to annotate these variants with functional information 
to gain a deeper understanding of their potential impact on the genome

Starting from the VCF file generated during the Detection step or provided by the user, 
**Musta** performs accurate and confident annotation by leveraging the power of two annotation tools:

- [VEP](https://www.ensembl.org/info/docs/tools/vep/index.html)
- [Funcotator](https://gatk.broadinstitute.org/hc/en-us/articles/360037224432-Funcotator). 

These annotation tools provide comprehensive information about the functional consequences of each variant, 
helping researchers understand the biological implications of the detected somatic mutations.

---

## Basic Usage
The command structure for invoking the Classification module in **Musta** is as follows: 

```shell 
musta classify [options]
```

By default, **Musta** uses VEP for variant annotation, but users can also choose to use Funcotator or a combination of both.

Here is a breakdown of the available options for the `musta classify` command:

**General Options:**
- `-h`, `--help`: Show the help message and exit.

**Required Options:**
- `--workdir PATH` (`-w PATH`): Specifies the destination folder for analysis. This is where the Snakemake pipeline, logs, and analysis outputs will be located.
- `--samples-file PATH` (`-s PATH`): Points to a YAML file listing the datasets you wish to analyze. This file provides information about the samples to be processed. For further details, please refer to the [dedicate section]({{site.url}}/pages/03.config/#input-datasets).
- `--reference-file PATH` (`-r PATH`): Identifies the path to the reference FASTA file, which is essential for the analysis. For further details, please refer to the [dedicate section]({{site.url}}/pages/03.config/#reference-file).
- `--data-source PATH` (`-ds PATH`): Specifies the path to a data source folder for Variant Annotations. To download data sources for VEP and Funcotator and organize the folders correctly, please refer to the [dedicate section]({{site.url}}/pages/03.config/#rclassfication-sources) 

**Optional Options:**
- `--ref-version {hg19, hg38}` or `-rf {hg19, hg38}`: Specifies the version of the Human Genome reference to use (e.g., hg19 or hg38). By default, this option is set to **hg19**
- `--cache-version CACHE_VERSION` or `-cv CACHE_VERSION`: Specifies the version of the offline cache to use with VEP (e.g., 75, 91, 102, 105, 106). Specifies the version of the offline cache to use with VEP. By default, this option is set to **106**
- `--tmpdir PATH` or `-t PATH`: Specifies the path to the temporary directory. Ensure that this directory has sufficient storage capacity to hold the intermediate files generated during the analysis.

**Variant Annotator Filtering Options:**
- `--also-funcotator` or `-af`: Runs GATK Funcotator in addition to VEP.
- `--only-funcotator` or `-of`: Runs GATK Funcotator exclusively without VEP.

**Additional Options:**
- `--force`, `-f`: Forces the re-creation of all output files.
- `--dryrun`, `-d`: Describes the workflow but does not execute it. This is useful for previewing the analysis steps before running them.

---

## Quick Start

1. **Obtain the "test-data-somatic" Dataset:**
   First, download the "test-data-somatic" dataset by following the instructions provided in the [dedicated section]({{site.url}}/pages/03.config/#demo-dataset) of this user guide. 

2. **Create an Input Samples YAML File:**
   Prepare an input `samples.yml` file, specifying your sample names, normal and tumor sample information, and paths to the corresponding BAM and VCF files. Here's an example YAML structure:

   ```yaml
   sample_A25:
       normal_sample_name:
           - N1
       tumor_sample_name:
           - A25
       normal_bam:
           - /path/to/test-data-somatic/data/bam/N1.chr22.bam
       tumor_bam:
           - /path/to/test-data-somatic/data/bam/A25.chr22.bam
       vcf:
           - /path/to/test-data-somatic/data/vcf/sample_A25.consensus.snvs.vcf.gz

   sample_B33:
       normal_sample_name:
           - N1
       tumor_sample_name:
           - B33
       normal_bam:
           - /path/to/test-data-somatic/data/bam/N1.chr22.bam
       tumor_bam:
           - /path/to/test-data-somatic/data/bam/B33.chr22.bam
       vcf:
           - /path/to/test-data-somatic/data/vcf/sample_B33.consensus.snvs.vcf.gz

   sample_C2:
       normal_sample_name:
           - N1
       tumor_sample_name:
           - C2
       normal_bam:
           - /path/to/test-data-somatic/data/bam/N1.chr22.bam
       tumor_bam:
           - /path/to/test-data-somatic/data/bam/C2.chr22.bam
       vcf:
           - /path/to/test-data-somatic/data/vcf/sample_C2.consensus.snvs.vcf.gz
   ```
   Modify it according to your specific requirements.

3. **Download Classification Sources:** Due to their large size and the complexity of reducing them to a demo version specific to chromosome 22, 
have not been included them in the "test-data-somatic" dataset.  Users are invited to independently download these resources by following the instructions provided in the [dedicated section]({{site.url}}/pages/03.config/#classification-sources). 

4. **Run Musta Classify with both Variant Annotator Enabled:** To execute the "Musta Classify" module with both variant annotators enabled, use the following command with the "test-data-somatic" dataset.

   ```shell
   musta classify \
   --workdir /path/to/workdir \
   --samples-file /path/to/samples.yml \
   --reference-file /path/to/test-data-somatic/resources/chr22.fa \
   --data-source /path/to/data-source \
   --also-funcotator
   ```
   
    Ensure you replace the /path/to/ placeholders with the actual file paths on your system.

These steps will help you quickly initiate somatic mutation classification using the “test-data-somatic” dataset within the Musta Detect module.

After initiating the execution of `musta classify`, you will see the Musta log unfolding on your screen:

```text
YYYY-MM-DD hh:mm:ss|INFO    |main |musta v1.0.0 - End-to-end pipeline to detect, classify and interpret mutations in cancer
YYYY-MM-DD hh:mm:ss|INFO    |main |Somatic Mutations Classification.    
Functional annotation of called somatic variants
VEP and/or GATK Funcotator
YYYY-MM-DD hh:mm:ss|INFO    |main |Reading configuration file
YYYY-MM-DD hh:mm:ss|INFO    |main |Setting paths
YYYY-MM-DD hh:mm:ss|INFO    |main |Deploying musta:v1.2.0.1 pipeline from https://github.com/solida-core/musta.git
YYYY-MM-DD hh:mm:ss|INFO    |main |Initializing  Config file
YYYY-MM-DD hh:mm:ss|INFO    |main |Initializing  Samples file
YYYY-MM-DD hh:mm:ss|INFO    |main |Running
YYYY-MM-DD hh:mm:ss|INFO    |main |Variant Annotation
YYYY-MM-DD hh:mm:ss|INFO    |main |Variant Annotator:  'vep'
Building DAG of jobs...
```
After a few minutes, you will see the last lines of the log, and the execution of Musta will conclude.
You can locate the resulting analysis in the following destination folders:

- **Logs**:    \<WORKDIR>/logs/<VARIANT ANNOTATOR>
- **Outputs**: \<WORKDIR\>/outputs/classification/<VARIANT ANNOTATOR>
- **Reports**: \<WORKDIR\>/outputs/classification/<VARIANT ANNOTATOR>/report.html
- **VCFs/MAFs**: \<WORKDIR\>/outputs/classification/results
- **Summary**: \<WORKDIR\>/outputs/classification/results/summary

These folders will contain the relevant data and reports generated during the `musta classify` execution.

---

## Exploring \<WORKDIR\> Folder
Let's take a look at the `<WORKDIR>` folder

```shell
.
├── benchmarks
│   ├── ...
│   └── classification
├── logs
│   ├── ...
│   └── classification
├── musta
│   ├── config
│   ├── LICENSE
│   ├── README.md
│   ├── resources
│   └── workflow
└── outputs
    ├── YYYYMMDD-hhmmss.report.html
    ├── ...
    └── classification
```

- **benchmarks**: This directory stores benchmark data and results, including performance metrics for individual variant annotators used in the classification module.
    ```shell
      ─ benchmarks
    │   ├── ...
    │   └── classification
    │       ├── vep
    │       │   ├── sample_A25.vep.txt
    │       │   ├── sample_B33.vep.txt
    │       │   ├── sample_C2.vep.txt
    │       │   ├── sample_A25.vep2maf.txt
    │       │   ├── sample_B33.vep2maf.txt
    │       │   └── sample_C2.vep2maf.txt
    │       └── funcotator
    │       │   ├── sample_A25.funcotator.vcf.txt
    │       │   ├── sample_B33.funcotator.vcf.txt
    │       │   ├── sample_C2.funcotator.vcf.txt
    │       │   ├── sample_A25.funcotator.vcf2maf.txt
    │       │   ├── sample_B33.funcotator.vcf2maf.txt
    │       │   └── sample_C2.funcotator.vcf2maf.txt
    ```

- **logs**: The logs directory stores various log files to keep track of Musta's activities.
  - `classification`: In this subdirectory, you'll find log files related to the classification module. These logs provide detailed information about the execution of various variant annotators.
  
    ```text
    ── logs
    │   ├── ...
    │   └── classification
    │       ├── vep
    │       │   ├── sample_A25.vep.log
    │       │   ├── sample_B33.vep.log
    │       │   ├── sample_C2.vep.log
    │       │   ├── sample_A25.vep2maf.log
    │       │   ├── sample_B33.vep2maf.log
    │       │   └── sample_C2.vep2maf.log
    │       └── funcotator
    │       │   ├── sample_A25.funcotator.vcf.log
    │       │   ├── sample_B33.funcotator.vcf.log
    │       │   ├── sample_C2.funcotator.vcf.log
    │       │   ├── sample_A25.funcotator.vcf2maf.log
    │       │   ├── sample_B33.funcotator.vcf2maf.log
    │       │   └── sample_C2.funcotator.vcf2maf.log
    ```

- **musta**: The directory houses the core of the Musta tool, including the Snakemake pipeline, which can be accessed from the official Musta GitHub repository: [https://github.com/solida-core/musta](https://github.com/solida-core/musta).

- **outputs**: The outputs directory  is designed to house the outcomes and data produced during the execution of Musta's classification module. It consists of the following subdirectories:
  - `YYMMDD-hhmmss.report.html`: Auto-generated, independent, detailed HTML reports that include runtime statistics, provenance information, workflow topology, and results.
  - `classification`: The directory houses the outcomes and data derived from the execution of Musta's detection module. Within this directory, users can explore a wealth of output files and data that are byproducts of the analysis. Specifically, you can find:
    - **VCF Files**: generated by each variant annotator enabled during the analysis. 
    - **MAF Files**: generated for each previous VCF file.
    - **HTML Reports**: Snakemake report HTML files. Users can access individual reports for each variant annotator.
    - **Statistics Files**:  Runtime statistics for each variant annotator.

    ```shell
    .
    ├── vep
    ├── funcotator
    ├── results
    │   ├── sample_A25.annotated.vep.vcf.gz
    │   ├── sample_A25.annotated.vep.vcf.gz.tbi
    │   ├── sample_A25.annotated.vep.maf
    │   ├── sample_A25.annotated.funcotator.vcf.gz
    │   ├── sample_A25.annotated.funcotator.vcf.gz.tbi
    │   ├── sample_A25.annotated.funcotator.maf
    │   ├── sample_B33....
    │   └── sample_C2....
    ```

---
### Summary Files

Within the `results` directory, 
a dedicated `summary` folder has been organized to compile and present essential aggregated data. 
This structured collection aids users in comprehending and interpreting key metrics and visual representations. 
Below is a detailed list of the files available in the `summary` folder:


**TODO**

---

## Usage Examples

Explore different scenarios for running **Musta** with various configurations using the `musta classify` command.

**1. Run Musta with Only VEP as Variant Annotator:**

Run Musta using VEP to annotate somatic variants:

```shell
musta classify \
--workdir /path/to/workdir \
--samples-file /path/to/samples.yml \
--reference-file /path/to/reference.fa \
--data-source /path/to/data_source
```

**2. Run Musta with Funcotator as Variant Annotator:**

Run Musta using GATK Funcotator to annotate somatic variants:

```shell
musta classify \
--workdir /path/to/workdir \
--samples-file /path/to/samples.yml \
--reference-file /path/to/reference.fa \
--data-source /path/to/data_source \
--only-funcotator
```

**3. Run Musta with Both VEP and Funcotator as Variant Annotators:**

Run Musta using both VEP and GATK Funcotator to annotate somatic variants:

```shell
musta classify \
--workdir /path/to/workdir \
--samples-file /path/to/samples.yml \
--reference-file /path/to/reference.fa \
--data-source /path/to/data_source \
--also-funcotator
```

Users can customize the Human Genome reference version and the offline VEP cache version using the options: 
`--ref-version {hg19, hg38}` and `--cache-version CACHE_VERSION`.


<p style="text-align:right"><a href="{{site.url}}{{page.url}}"><strong>Go Up</strong><span class="fa fa-fw fa-arrow-up"></span></a></p>
---

